rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,710 @@
+declarative macros, it represents a section of code that may be harder to understand and modify than
+if we used a more brute force approach.
+
+## Prior Art
+
+Notably, and perhaps most obviously, Cue itself can be used for data validation purposes. There are
+also other projects that slot themselves into the same general domain, such as [OPA/Rego][opa_rego]
+and [CDDL][cddl].
+
+The primary problem with all of these solutions is that they're all custom languages, with far less
+cross-platform support, and a far steeper learning curve. Even though our approach generally
+concerns itself with obfuscating the schema tool itself, and focusing on making it trivially to
+generate the schema from annotated Rust types/fields, eventually the rubber must meet the road, and
+this is where these other tools would fall down for our case and become very hard to wield
+correctly.
+
+## Alternatives
+
+While `schemars` itself has support for annotating Rust types in almost the exact same way as we've
+proposed above, it lacks a few features necessary for our use case:
+
+- no support for the serde `alias` attribute feature
+- no support for defining generic metadata for a type/field and exposing it in the schema
+- no mechanism to override constraints for a field which already defines its own constraints at the
+  type level
+
+The lack of these features make it much harder to correctly generate our schema, as we would still
+be required to do the minimum amount of work to support defining custom attributes for the missing
+features, and the work to support at least one attribute is 90% of the work to support two, or
+three, or more, custom attributes. It would also mean that developers would need to figure out
+whether or not `schemars` provided a certain behavior via its attributes, versus being able to only
+need to remember how to get to the documentation for the our proposed internal approach.
+
+## Outstanding Questions
+
+- Can the proposed solution work seamlessly for all types that we use in our configuration? (For
+  example, `Option<T>` or any other type that needs to partially forward its declaration to the
+  underlying `T`)
+- Can downstream consumers extract the necessary information purely from the specified JSONSchema
+  fields and any additional custom metadata as string key/value pairs?  Do they need a richer
+  representation of custom metadata?
+- Is there a better way, or a way at all, to handle overriding subfields at compile-time?
+  (Procedural macros operate on the Rust AST, so for example, we cannot annotate a field that points
+  to type B and reference a field that only exists on B, so we have to generate code such that it
+  eventually fails at run-time.)
+- How can we provide logical constraints between disparate components? For example, using disk
+  buffers implies needing a global data directory set, but there's not really a way in JSONSchema to
+  add a conditional constraint to a remote definition/schema.
```

Comment:
It's not possible for `data_dir` to be empty, right? So your example should be impossible. Regardless, JSON schema has some facilities for conditional sub-schemas, [`dependentRequired` comes to mind](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.5.4)

Reply:
`dependentRequired` can't cross "modules", as far as I can tell, so it's only useful if you're dealing with properties in the same object.

Beyond that, though, what you're saying makes sense.